KOGITO-7066: SWF - Expose Stunner API through multiplying architecture

romartin opened the pull request:
See [KOGITO-7066](https://issues.redhat.com/browse/KOGITO-7066)

**How to test the API from the UI**

Notice the API is implemented at _combined editor_ level, so it is accessible to most of the channels, except VSCode extension. Here will share some instructions about how to test it by using the _web-tools_ channel.

1. Download and extract the `build_artifacts.zip` , from this PR (_Checks_ tab)
2. Run a web server in the root folder: `build_artifacts/packages/serverless-logic-web-tools/dist`. _I personally use the Live Server extension in VSCode, I just open the dist/ folder and run the server._
4. Open `index.html` in your browser (located at root `build_artifacts/packages/serverless-logic-web-tools/dist/index.html`)
5. Once having the web-tools UI available, just create or open a workflow
6. Open the chrome _Console_ tab, from the web tools
7. TIP: Some errors are displayed, in order not to make noise, please filter our errors  from the console, Unselect the `errors` item here:
![stunner-envelope-api_filterout_errors](https://github.com/kiegroup/kie-tools/assets/4602417/d2a79e44-eb84-43c4-be14-548ffdddecbb)
8. In the console, choose the right context by selecting the `serverless-workflow-combined-editor  iframe `.  See:
![stunner-envelope-api_frame_selection](https://github.com/kiegroup/kie-tools/assets/4602417/35fb7ae9-3db3-4dc4-b7a2-b72f883995a1)



**Examples**
Notice the API is documented at https://github.com/kiegroup/kie-tools/pull/1589/files#diff-292fd900d04b40fbd39f26fefa506b17357917982fce264719152ed0ee62d913R23
From this point, you can use the API in the console, here are some examples:
**a) Obtain the selected node UIUID:**
a.1) Select a node from the UI, by clicking on any state
a.2) `editor.session.getSelectedElementUUID().then(uuid => console.log(uuid))`

**b) Change colors for an state**
b.1) Ensure nothing is selected on the UI
b.2) `editor.canvas.setBackgroundColor('_0757D1D8-C613-4B1A-955E-300BB2818D92', 'grey').then(done => {})`
b.3) `editor.canvas.setBorderColor('_0757D1D8-C613-4B1A-955E-300BB2818D92', 'black').then(done => {})`
b.4) `editor.canvas.draw().then(done => {})`
b.5) See results:
![stunner-envelope-api_canvas_example](https://github.com/kiegroup/kie-tools/assets/4602417/4956e5f8-59f8-40bd-8314-0635c8164754)

AjayJagan commented:
awesome!! 👍

AjayJagan commented on packages/serverless-workflow-combined-editor/src/editor/ServerlessWorkflowCombinedEditor.tsx:
```diff
@@ -400,6 +413,11 @@ const RefForwardingServerlessWorkflowCombinedEditor: ForwardRefRenderFunction<
     )
   );
 
+  window.editor = new SwfStunnerEditor(
+    diagramEditor?.getEnvelopeServer()
+      .envelopeApi as unknown as MessageBusClientApi<ServerlessWorkflowDiagramEditorEnvelopeApi>
+  );
+
```
```suggestion
    window.editor = useMemo(() => new SwfStunnerEditor(
    diagramEditor?.getEnvelopeServer()
      .envelopeApi as unknown as MessageBusClientApi<ServerlessWorkflowDiagramEditorEnvelopeApi>
  ), [diagramEditor])
```

AjayJagan replied:
maybe put it inside useMemo?

romartin replied:
sure Ajay, working on it, thanks!

AjayJagan reviewed (approved):
Thanks for the great work Roger!! Everything works on my side.

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  /**
+   * Obtain the shape's background (fill) color.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @returns Asyncrhonously returns the background color (name or RGB formatted value) for the shape.
```
```suggestion
   * @returns Asynchronously returns the background color (name or RGB formatted value) for the shape.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   * Sets the shape's background (fill) color.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @param backgroundColor The color name or RGB value for the background color.
+   * @returns Asyncrhonously notifies when the background color is being changed. Notice you may call draw once all attributes have been changed.
```
```suggestion
   * @returns Asynchronously notifies when the background color is being changed. Notice you may need to call a draw function once all attributes have been changed.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  setBackgroundColor(uuid: string, backgroundColor: string): Promise<void>;
+  /**
+   * Obtain the shape's border (stroke) color.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @returns Asyncrhonously returns the border color (name or RGB formatted value) for the shape.
```
```suggestion
   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
   * @returns Asynchronously returns the border color (name or RGB formatted value) for the shape.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  getBorderColor(uuid: string): Promise<string>;
+  /**
+   * Sets the shape's border (stroke) color.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @param backgroundColor The color name or RGB value for the border color.
+   * @returns Asyncrhonously notifies when the border color is being changed. Notice you may call draw once all attributes have been changed.
```
```suggestion
   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
   * @param backgroundColor The color name or RGB value for the border color.
   * @returns Asynchronously notifies when the border color is being changed. Notice you may need to call a draw function once all attributes have been changed.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  setBorderColor(uuid: string, backgroundColor: string): Promise<void>;
+  /**
+   * Obtain the shape's location relative to the parent, if any.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @returns Asyncrhonously returns the location (X and Y coordinates in a 2D space) for the given shape.
```
```suggestion
   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
   * @returns Asynchronously returns the location (X and Y coordinates in a 2D space) for the given shape.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  getLocation(uuid: string): Promise<number[]>;
+  /**
+   * Obtain the shape's location relative to the viewport.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @returns Asyncrhonously returns the absolute location (X and Y coordinates in a 2D space) for the given shape.
```
```suggestion
   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
   * @returns Asynchronously returns the absolute location (X and Y coordinates in a 2D space) for the given shape.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  setBorderColor(uuid: string, backgroundColor: string): Promise<void>;
+  /**
+   * Obtain the shape's location relative to the parent, if any.
```
```suggestion
   * Obtains the shape's location relative to the parent, if any.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  getLocation(uuid: string): Promise<number[]>;
+  /**
+   * Obtain the shape's location relative to the viewport.
```
```suggestion
   * Obtains the shape's location relative to the viewport.
```

LuboTerifaj replied:
```suggestion
   * Obtains the shape's location relative to the viewport.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  getAbsoluteLocation(uuid: string): Promise<number[]>;
+  /**
+   * Compute the shape's dimensions.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @returns Asyncrhonously returns the width and height values (in PX units) for the given shape.
```
```suggestion
   * Computes the shape's dimensions.
   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
   * @returns Asynchronously returns the width and height values (in PX units) for the given shape.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  /**
+   * Changes the viewport accordingly to render a shape in its center.
+   * Only applies when shapes is not visible at all, in the viewport area.
+   * @param uuid The shape's identifier. It uses to match with the graph element's identifier.
+   * @returns Asyncrhonously notifies when operations are being completed.
```
```suggestion
   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
   * @returns Asynchronously notifies when operations are being completed.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  center(uuid: string): Promise<void>;
+  /**
+   * It performs the screen rendering for the changed objects or shapes' attributes in the editor's view.
+   * @returns Asyncrhonously notifies when operations are being completed.
```
```suggestion
   * @returns Asynchronously notifies when operations are being completed.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  /**
+   * It return the name for a Node.
+   * @param uuid The node's identifier (UUID).
+   * @returns Asyncrhonously returns the name.
```
```suggestion
   * @returns Asynchronously returns the name.
```

LuboTerifaj commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  getNodeName(uuid: string): Promise<string>;
+  /**
+   * It returns the graph element (Node, Edge) identifier (UUID) for the selected shape in the editor's view.
+   * @returns Asyncrhonously returns the element's identifier (UUID).
```
```suggestion
   * @returns Asynchronously returns the element's identifier (UUID).
```

LuboTerifaj reviewed (commented):
Hello @romartin ,

Could you please update the code documentation? See my inline comments.

The testing part of my review will follow.

Thank you!

romartin commented on packages/serverless-workflow-diagram-editor/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/api/JsWindow.java:
```diff
@@ -10,6 +10,7 @@ public class JsWindow {
     @JsProperty
     public static JsStunnerEditor editor;
 
+    @Deprecated // Use editor.canvas instead
```
Hey @LuboTerifaj , yeah the idea is to finally drop this, but I am aware of a few other places in our repositories that are still using this, so I prefer to deprecate first and progressively, do the changes later on on all third party usages.
This way, there exist this ticket that describes next steps:
https://issues.redhat.com/browse/KOGITO-9146 
Does it makes sense to you? Thx

LuboTerifaj replied:
Yes, it makes sense. Thank you for clarification.

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/StunnerEditorEnvelopeAPI.ts:
```diff
@@ -0,0 +1,106 @@
+export class Definition {
+  /**
+   * The domain model object's identifier.
+   * It uses to match with FQCN of the objec'ts class.
```
done

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  session: SwfStunnerEditoSession;
+  /**
+   * The diagram editor's canvas object. It allows to get or update shape attributes' values,
+   * as well other methods related to the management of the editor's view.
```
done

romartin commented:
Hey @LuboTerifaj , thanks for the review on the API docs, is great! I am agree, also applied all changes requested from you :+1: 

Also, on this PR's description you can find instructions for quickly testing the API usage, manually. 

Please let me know if any further concerns or issues. Thanks!

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   * @param backgroundColor The color name or RGB value for the border color.
+   * @returns Asynchronously notifies when the border color is being changed. Notice you may need to call a draw function once all attributes have been changed.
+   */
+  setBorderColor(uuid: string, backgroundColor: string): Promise<void>;
```
Fixed, thanks

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+  /**
+   * Sets the shape's border (stroke) color.
+   * @param uuid The shape's identifier. It is used to match with the graph element's identifier.
+   * @param backgroundColor The color name or RGB value for the border color.
```
Fixed, thanks

romartin commented:
Hey @LuboTerifaj 

Thanks for the review, again! Here are the updates:

- Fixed all typos in docs, as you suggested
- Fixed the API argument for the method `getNodeName`. Sorry, I did a mistake in the TS API and it was assuming an `string`, but in fact the (internal) stunner API defines the argument as a `Node`. Anyway, fixed, sorry for the inconveniences. You can now do, once having some node selected, this: `editor.session.getSelectedNode().then(node => editor.session.getNodeName(node)).then(name => console.log(name))`
- About the `getSelectedDefinition`, it works for me! Did you maybe tested once having nothing selected on the UI? 
- Just as a comment about the use of the API: Notice that if any promise rejects or any exception is thrown at some point when using the API , it is ok!! we expect valid ids and instances, so it is up to the third parties to do so, and capture any runtime exception, if any. 

This way, everything fixed and pushed! Your turn again, thanks!!! :+1:

LuboTerifaj reviewed (changes requested):
Hello @romartin ,

there are still a few updates needed in the code to have the same documentation style of the API.

The function with changed parameter works as expected.

I couldn't have tested the following one properly:

- For `editor.session.getSelectedDefinition().then(definition => console.log(definition))` I am still getting "Promise rejected". I don't know how to check the behaviour of this function properly. Could you please share with me what to expect here and how the definition object can be used?

Regarding other channels, I didn't mean to test every single function, but to check if the API is accessible in all channels. However I see your point that it is exposed to SWF Combined Editor, so no specific channels need to be tested.

Thank you!

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  getAllNodesUUID(): Promise<string[]>;
+  /**
+   * It returns the Edge for the given identifier (UUID).
```
done, thx

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  getEdgeByUUID(uuid: string): Promise<Edge>;
+  /**
+   * It returns the Node for the given identifier (UUID).
```
done, thx

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  getDefinitionByElementUUID(uuid: string): Promise<Object>;
+  /**
+   * It returns the Node which has a given name.
```
done, thx

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/SwfStunnerEditorAPI.ts:
```diff
@@ -0,0 +1,182 @@
+   */
+  getNodeByName(name: string): Promise<Node>;
+  /**
+   * It returns the name for a Node.
```
done, thx

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/StunnerEditorEnvelopeAPI.ts:
```diff
@@ -0,0 +1,106 @@
+  id: string;
+  /**
+   * The domain model object's name.
+   * It returns the value for the field that represents the name for a graph element.
```
I have rephrased a little bit trying to make it more comprehensible.

LuboTerifaj replied:
I meant that since this is not a function, there shouldn't be "It returns". The description was good enough.

romartin commented:
Hey @LuboTerifaj 

Thanks for the suggestions , changes applied!

Also, about the `getSelectedDefinition` and `getDefinitionByElementUUID` => The idea behind this methods is to obtain the `Definition` (which is the bean/object behind a node) for, either the actually selected node in the UI, or for the node given by an UUID. This way, we are decoupling the graph model, from the domain model objects. Usually when playing with this API it deals with graph objects, so in the envelope bus we are only transferring the node but not all the information about its domain model (for example, we are not transferring all the inject state object and fields). 

It will be the end user who explicitly has to call any of those methods to obtain (transfer) the objects for a given node:

- for example, select some inject state node in the UI, and then call the `getSelectedDefinition()` method => It will return the inject state object that is contained by the selected node
- for example, obtain the UUID for some node in the canvas which represents an inject state, and then call `getDefinitionByElementUUID(uuid)` => It will return the inject state object that is contained by the node which has the given uuid

LuboTerifaj reviewed (approved):
Hello @romartin ,

please see my last inline comment.

Regarding the issue I reported above, it is present in this PR. I have reported it in separate jira: https://issues.redhat.com/browse/KOGITO-9201 _SWF - getSelectedDefinition API function doesn't translate all fields_

Thanks!

romartin commented on packages/serverless-workflow-diagram-editor-envelope/src/api/StunnerEditorEnvelopeAPI.ts:
```diff
@@ -0,0 +1,107 @@
+  id: string;
+  /**
+   * The domain model object's name.
+   * It returns the value for the name field for the object that this definition represents (eg: the name of an InjectState, or an OperationState).
```
fixed, thanks!
